Draw "available job" pointers before "active job" pointers

Jamibaraki opened the pull request:
Ref #4436

This PR swaps draw code for the available and active missions, leaving the available missions visible when more than 12 missions are active in a system. You can see a hint of blue behind them as they are drawn over the top of them.

I've tested this with a save file with many missions available, though not with special missions.

MCOfficer commented:
please, give your PR's descriptive titles. nobody knows by heart what 4436 is.

Jamibaraki commented:
Sure, I'll try and change it to do it that way.

tehhowch commented:
I agree with Peter.

The special job system pointer should probably be drawn first, and then the remaining slots should be filled (possibly proportionally) by active and available jobs with the respective colorings.

We can consider a tooltip when hovering over the pointer ring that displays the job category count.

Jamibaraki commented:
OK, I've updated this, replacing the <map angles> with a map of integers to count the missions per system, and only drawing a max of 12.

As suggested, I've also moved the special system code to draw first. I'm not sure how to test the special systems in the game itself though.

The code at line 264 includes a temporary int variable to feed into the function, which is a little clunky, but it was doing the same thing with a temporary angle before.

Please let me know if there are any issues with it, then after that I'll look at adding a tooltip.

petervdmeer commented on source/MapPanel.cpp:
```diff
@@ -1143,16 +1145,21 @@ void MapPanel::DrawTooltips()
+	angle += Angle(30.*systemCount);
+	if(systemCount>12) {
+        return;
+    }
```
There is an alignment/indentation issue here with tabs and spaces.

petervdmeer commented on source/MapPanel.cpp:
```diff
@@ -1143,16 +1145,21 @@ void MapPanel::DrawTooltips()
-	angle += Angle(30.);
+	Angle angle = Angle();
+	++systemCount;
+	angle += Angle(30.*systemCount);
```
Minor detail, but I think it's possible to construct the Angle in one go:
`Angle angle = Angle(30.*systemCount);`
Also for the special system above.

Jamibaraki commented:
OK - updated. Does that fix it correctly?

petervdmeer commented on source/MapPanel.cpp:
```diff
@@ -1143,16 +1144,20 @@ void MapPanel::DrawTooltips()
+	++systemCount;
+	Angle angle = Angle(30.*systemCount);
+	if(systemCount>12) {
+    	return;
```
There still is an issue with spaces instead of tabs here.

Jamibaraki replied:
Thanks for spotting it! Is it OK now?

petervdmeer replied:
Yes, it is OK now.

Jamibaraki commented:
Here are screenshots of the first two requested situations (1 available + 13 active, and 1 active and 13 available). Next I'll try to figure out how to get a save with the special mission active as well. If anyone would happen to have a save file representing that specific scenario, feel free to share :-)

![13active1available](https://user-images.githubusercontent.com/980528/63773057-5b9ca700-c915-11e9-9bb5-33e4493888fe.png)
![13available1active](https://user-images.githubusercontent.com/980528/63773944-2ee98f00-c917-11e9-8cc2-271b07f096fb.png)

Jamibaraki commented:
Here are two more screenshots taken after being assigned a special mission when landing and selecting spaceport, then, after quitting and adding all the extra missions at that location in the save file, reloading and viewing the map screen.

![13active1availableWithSpecial](https://user-images.githubusercontent.com/980528/64155440-c0a05180-ce6d-11e9-8475-e082ae332336.png)
![13available1activeWithSpecial](https://user-images.githubusercontent.com/980528/64155452-c7c75f80-ce6d-11e9-8dbf-c3b699d07f43.png)

tehhowch commented:
Your last two screenshots don't show the special system pointer (view the map while being offered the spaceport mission, not after accepting it).

Your "many available + 1 active" screenshot doesn't show the blue pointer for an active mission

Jamibaraki commented:
Thanks. I was wondering how to actually view the special system pointer on the map. The trouble now is that I'm not sure how to get it to appear in a specific system for the purpose of generating a screenshot. I have a saved game where I can fly to a system that has a mission available, but it isn't always generated for the same location.

I'm creating the 13+1 active/available scenarios by modifying the save game file before or after flying to this system. I could alter the code to generate a special mission in the correct system, but then the screenshots wouldn't be from the actual code in the pull request.

Moving on to the second point about not showing the blue pointers - the goal was to fix the orange pointers not showing in the inverse situation, so this fix wouldn't be expected to show the blue pointer. 

I'm not sure if that situation will arise in game, but if so, we could switch it to draw whichever mission type has the lowest count first, perhaps? Or put a plus next to the systems with hidden missions?

What do you think?

petervdmeer commented:
I updated this PR with a merge from master and with the functionality to show both available jobs and accepted jobs when the total amount for some system is more than 12.

Tested with the savegame I provided on #4436:
- Nothing accepted:
![nothing_accepted](https://user-images.githubusercontent.com/35403542/161402376-b510ba90-05f7-491b-9104-3afb4038509a.png)
- Special marker (and blinking destination): 
![marker_and_blinking](https://user-images.githubusercontent.com/35403542/161402355-4e6ba11e-d1d1-47cc-b4c3-9b453ea7a960.png)
- About half accepted:
![about_half_accepted](https://user-images.githubusercontent.com/35403542/161402391-82f73c2f-29c1-49f2-abf0-2d093a968e36.png)
- Most accepted:
![most_accepted](https://user-images.githubusercontent.com/35403542/161402405-34fa2310-73ab-4772-9600-5b66aa14b1d5.png)
- All accepted:
![all_accepted](https://user-images.githubusercontent.com/35403542/161402409-6e484dc1-6449-451f-98c6-c46fa3007355.png)

quyykk commented on source/MapPanel.cpp:
```diff
@@ -1152,17 +1153,37 @@ void MapPanel::DrawMissions()
+	{
+		// The special system pointer is larger than the others.
+		pointerCount[specialSystem] += 1;
+		Angle a = Angle(30.*pointerCount[specialSystem] );
```
```suggestion
		Angle a = Angle(30. * pointerCount[specialSystem]);
```

quyykk commented on source/MapPanel.cpp:
```diff
@@ -1152,17 +1153,37 @@ void MapPanel::DrawMissions()
 		const System *system = mission.Destination()->GetSystem();
-		DrawPointer(system, angle[system], mission.CanAccept(player) ? availableColor : unavailableColor);
+		if(mission.CanAccept(player))
+			++(availableCount[system].first);
```
```suggestion
			++availableCount[system].first;
```

? There seems to be examples with and without throughout the code. We should really decide which one. My vote is without.

quyykk commented on source/MapPanel.cpp:
```diff
@@ -1171,20 +1192,22 @@ void MapPanel::DrawMissions()
-		PointerShader::Draw(pos, a.Unit(), 11.5f, 21.5f, -6.f, specialColor);
+		auto &system = it.first;
+		auto &available = it.second;
+		for(int i=0; i < available.first; i++)
```
```suggestion
		for(int i = 0; i < available.first; ++i)
```

quyykk commented on source/MapPanel.cpp:
```diff
@@ -1171,20 +1192,22 @@ void MapPanel::DrawMissions()
+		auto &available = it.second;
+		for(int i=0; i < available.first; i++)
+			DrawPointer(system, pointerCount[system], availableColor);
+		for(int i=0; i < available.second; i++)
```
```suggestion
		for(int i = 0; i < available.second; ++i)
```

quyykk commented on source/MapPanel.cpp:
```diff
@@ -1254,16 +1277,20 @@ void MapPanel::DrawTooltips()
-	angle += Angle(30.);
+	++systemCount;
+	Angle angle = Angle(30.*systemCount);
+	if(systemCount>12) {
+		return;
+	}
```
```suggestion
	if(systemCount > 12)
		return;
```

quyykk commented on source/MapPanel.cpp:
```diff
@@ -1254,16 +1277,20 @@ void MapPanel::DrawTooltips()
 {
-	angle += Angle(30.);
+	++systemCount;
+	Angle angle = Angle(30.*systemCount);
```
```suggestion
	Angle angle = Angle(30. * systemCount);
```

quyykk commented on source/MapPanel.cpp:
```diff
@@ -1171,20 +1192,22 @@ void MapPanel::DrawMissions()
-		Point pos = Zoom() * (specialSystem->Position() + center);
-		PointerShader::Draw(pos, a.Unit(), 20.f, 27.f, -4.f, black);
-		PointerShader::Draw(pos, a.Unit(), 11.5f, 21.5f, -6.f, specialColor);
+		auto &system = it.first;
```
```suggestion
		const auto &system = it.first;
```

quyykk commented on source/MapPanel.cpp:
```diff
@@ -1171,20 +1192,22 @@ void MapPanel::DrawMissions()
-		PointerShader::Draw(pos, a.Unit(), 20.f, 27.f, -4.f, black);
-		PointerShader::Draw(pos, a.Unit(), 11.5f, 21.5f, -6.f, specialColor);
+		auto &system = it.first;
+		auto &available = it.second;
```
```suggestion
		const auto &available = it.second;
```

quyykk commented on source/MapPanel.cpp:
```diff
@@ -1152,17 +1153,37 @@ void MapPanel::DrawMissions()
+	if(specialSystem)
+	{
+		// The special system pointer is larger than the others.
+		pointerCount[specialSystem] += 1;
```
```suggestion
		++pointerCount[specialSystem];
```

tehhowch commented on source/MapPanel.cpp:
```diff
@@ -1171,20 +1192,22 @@ void MapPanel::DrawMissions()
-		PointerShader::Draw(pos, a.Unit(), 11.5f, 21.5f, -6.f, specialColor);
+		const auto &system = it.first;
+		const auto &available = it.second;
+		for(int i = 0; i < available.first; ++i)
+			DrawPointer(system, pointerCount[system], availableColor);
+		for(int i = 0; i < available.second; ++i)
+			DrawPointer(system, pointerCount[system], unavailableColor);
```
Hoist the map indexing out of the loop.

petervdmeer commented on source/MapPanel.cpp:
```diff
@@ -1143,7 +1143,8 @@ void MapPanel::DrawNames()
 	// Draw a pointer for each active or available mission.
-	map<const System *, Angle> angle;
+	map<const System *, int> pointerCount;
+	map<const System *, pair<int, int>> availableCount;
```
I added a file-local struct.

quyykk commented on source/MapPanel.cpp:
```diff
@@ -59,6 +59,15 @@ namespace {
 
+	// Struct to track per system how many pointers are drawn and still
+	// need to be drawn.
+	struct PointerDrawCount{
+		// Amount of systems already drawn.
```
```suggestion
	class PointerDrawCount {
	public:
		// Amount of systems already drawn.
```

MZ doesn't like structs.

tehhowch replied:
I'm fine with their use when it is to provide named fields to what would otherwise be a pair or a tuple.

quyykk replied:
Yeah same. We'll need to update the style guide at some point.

quyykk commented on source/MapPanel.cpp:
```diff
@@ -1152,17 +1161,38 @@ void MapPanel::DrawMissions()
+		// Reserve a maximum of 6 slots for available missions.
+		auto &&it = pointerCount[system];
+		int reserved = min(6, it.available + it.unavailable);
+		if(it.drawn >= 12 - reserved)
```
Make the 12 a global constant with a cool name (since it's used in two places). Maybe also for the `30.` (degrees), or we calculate the degree from the constant if that makes sense.

tehhowch commented on source/MapPanel.cpp:
```diff
@@ -1254,16 +1286,19 @@ void MapPanel::DrawTooltips()
+void MapPanel::DrawPointer(Point position, int &systemCount, const Color &color, bool drawBack, bool bigger)
 {
-	angle += Angle(30.);
+	++systemCount;
+	Angle angle = Angle(30. * systemCount);
+	if(systemCount > 12)
+		return;
```
Do less work:

```suggestion
	if(++systemCount > 12)
		return;
	Angle angle = Angle(30. * systemCount);
```

tehhowch commented on source/MapPanel.cpp:
```diff
@@ -59,6 +59,15 @@ namespace {
+	// need to be drawn.
+	struct PointerDrawCount{
+		// Amount of systems already drawn.
+		int drawn = 0;
+		int available = 0;
+		int unavailable = 0;
```
Probably these should be `unsigned`

quyykk commented on source/MapPanel.cpp:
```diff
@@ -58,6 +58,17 @@ using namespace std;
+
+	// Struct to track per system how many pointers are drawn and still
+	// need to be drawn.
+	struct PointerDrawCount{
```
```suggestion
	struct PointerDrawCount {
```

quyykk commented on source/MapPanel.cpp:
```diff
@@ -320,9 +331,12 @@ void MapPanel::DrawMiniMap(const PlayerInfo &player, float alpha, const System *
 		}
 
-		Angle angle;
+		unsigned int missionCounter = 0;
```
I believe the style is to use plain `unsigned` over `unsigned int`.

petervdmeer replied:
Yes, that seems to be the common style; I changed them to `unsigned`.
Can we automatically check for this in the coding style checker? (Which I believe you were working on?)

quyykk replied:
No unfortunately not.

petervdmeer commented on source/MapPanel.cpp:
```diff
@@ -58,6 +58,17 @@ using namespace std;
 namespace {
 	const std::string SHOW_ESCORT_SYSTEMS = "Show escort systems on map";
 	const std::string SHOW_STORED_OUTFITS = "Show stored outfits on map";
+	const unsigned int MAX_MISSION_POINTERS_DRAWN = 12;
```
```suggestion
	const unsigned MAX_MISSION_POINTERS_DRAWN = 12;
```

petervdmeer commented on source/MapPanel.cpp:
```diff
@@ -58,6 +58,17 @@ using namespace std;
+	// need to be drawn.
+	struct PointerDrawCount{
+		// Amount of systems already drawn.
+		unsigned int drawn = 0;
+		unsigned int available = 0;
+		unsigned int unavailable = 0;
```
```suggestion
		unsigned drawn = 0;
		unsigned available = 0;
		unsigned unavailable = 0;
```

petervdmeer commented on source/MapPanel.cpp:
```diff
@@ -1171,20 +1216,22 @@ void MapPanel::DrawMissions()
-		PointerShader::Draw(pos, a.Unit(), 11.5f, 21.5f, -6.f, specialColor);
+		const auto &system = it.first;
+		auto &&counters = it.second;
+		for(unsigned int i = 0; i < counters.available; ++i)
```
```suggestion
		for(unsigned i = 0; i < counters.available; ++i)
```

petervdmeer commented on source/MapPanel.cpp:
```diff
@@ -1254,16 +1301,18 @@ void MapPanel::DrawTooltips()
 
 
-void MapPanel::DrawPointer(const System *system, Angle &angle, const Color &color, bool bigger)
+void MapPanel::DrawPointer(const System *system, unsigned int &systemCount, const Color &color, bool bigger)
```
```suggestion
void MapPanel::DrawPointer(const System *system, unsigned &systemCount, const Color &color, bool bigger)
```

petervdmeer commented on source/MapPanel.cpp:
```diff
@@ -1254,16 +1301,18 @@ void MapPanel::DrawTooltips()
 
 
-void MapPanel::DrawPointer(Point position, Angle &angle, const Color &color, bool drawBack, bool bigger)
+void MapPanel::DrawPointer(Point position, unsigned int &systemCount, const Color &color, bool drawBack, bool bigger)
```
```suggestion
void MapPanel::DrawPointer(Point position, unsigned &systemCount, const Color &color, bool drawBack, bool bigger)
```

petervdmeer commented on source/MapPanel.h:
```diff
@@ -163,8 +163,8 @@ class MapPanel : public Panel {
 	void DrawTooltips();
-	void DrawPointer(const System *system, Angle &angle, const Color &color, bool bigger = false);
-	static void DrawPointer(Point position, Angle &angle, const Color &color, bool drawBack = true, bool bigger = false);
+	void DrawPointer(const System *system, unsigned int &systemCount, const Color &color, bool bigger = false);
+	static void DrawPointer(Point position, unsigned int &systemCount, const Color &color, bool drawBack = true, bool bigger = false);
```
```suggestion
	void DrawPointer(const System *system, unsigned &systemCount, const Color &color, bool bigger = false);
	static void DrawPointer(Point position, unsigned &systemCount, const Color &color, bool drawBack = true, bool bigger = false);
```

petervdmeer commented:
All comments seem to have been processed; ready for merging.
